### PR TITLE
Data structure: use real arrays (#62)

### PR DESCRIPTION
**Pull request by Artoria2e5**
#62

### CODE REVIEW

**Comment by Arthur2e5** [comment]
Bai, try to run a basic self-build check..

File: build/20-ruby.sh [hunk 1]
```diff
@@ -4,14 +4,13 @@
...
 }
 
 build_ruby_build(){
-  GEMDIR="$(ruby -e'puts Gem.default_dir')"
-  gem install --ignore-dependencies --no-user-install \
-    -i "$PKGDIR/$GEMDIR" -n "$PKGDIR/usr/bin" $PKGNAME-$PKGVER.gem
-  rm "$PKGDIR/$GEMDIR/cache/$PKGNAME-$PKGVER.gem"
+	GEMDIR="$(ruby -e'puts Gem.default_dir')"
+	gem install --ignore-dependencies --no-user-install -i "$PKGDIR/$GEMDIR" -n "$PKGDIR/usr/bin" "$PKGNAME-$PKGVER.gem"
+	rm -f -- "$PKGDIR/$GEMDIR/cache/$PKGNAME-$PKGVER.gem"
```
Comment (Arthur2e5) [inline build/20-ruby.sh]: Should I simply nuke all the cache?

File: pm/dpkg/pack [hunk 2]
```diff
@@ -12,43 +12,40 @@ dpkgctrl(){
...
-			if ((instSIZE < 33552000)); then
-				DPKGDEBCOMP+="-z8"
-			else DPKGDEBCOMP+="-z9"
-			fi
-		fi
-	else
-		DPKGDEBCOMP+="-z6"
-	fi
-}
+if [[ " ${DPKGDEBCOMP[*]" == *' '-z* ]]; then
```
Comment (Arthur2e5) [inline pm/dpkg/pack]: wtf

**Comment by Arthur2e5** [comment]
rebased. still recommending `pull --rebase=preserve origin data`.

**Comment by Arthur2e5** [comment]
Well, OK, f that, I forgot to check the args in arch/_common.

**Comment by Arthur2e5** [comment]
Since the flags are exported into ENV in the traditional sloppy way anyway, there is no point in array-izing it for now.

**Comment by lmy441900** [comment]
How is this PR now?

**Comment by Artoria2e5** [comment]
Changes required for merging should be still obvious.

**Comment by Artoria2e5** [comment]
lmao this isn't going anywhere. Anyways, just came up with these snippets:
```
is_array() { [[ $(declare -p $1) == 'declare -'[aA]* ]]; }
to_array() {
  declare -n ref=$1
  mapfile -d ' ' -t ABDEPENDS < <(printf %s "$ref")
}
check_type_array() {
  if ! is_array "$1"; then
    ab_warn "$1 should be an array now! splitting."
    to_array "$1"
  fi
}
```

**Comment by MingcongBai** [comment]
This has stalled for way too long and requires quite a bit of work still. Let's resurrect this when we feel up for it.

**Pull request closed without merging**